agent/service_manager.go

```diff
@@ -0,0 +1,237 @@
+package agent
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/cache"
+	cachetype "github.com/hashicorp/consul/agent/cache-types"
+	"github.com/hashicorp/consul/agent/structs"
+	"golang.org/x/net/context"
+)
+
+// The ServiceManager is a layer for service registration in between the agent
+// and the local state. Any services must be registered with the ServiceManager,
+// which then maintains a long-running watch of any globally-set service or proxy
+// configuration that applies to the service in order to register the final, merged
+// service configuration locally in the agent state.
+type ServiceManager struct {
+	services map[string]*serviceConfigWatch
+	agent    *Agent
+
+	sync.Mutex
```

Comment:
nit: Is it desirable that Lock/Unlock are promoted to public methods that external callers can make? They seem like an implementation detail here. Not a huge deal but my preference these days is to keep a lock private unless you expect external code to actually manipulate the Lock and the protected resource - in this case they can't because all the other state is private anyway. It adds a few chars to each lock call but seems cleaner to me.